src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -723,13 +724,19 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             !runPaths = Paths.runPath root (uniqueToRunNumber n)
         traceWith tr $ AtLevel ln $
           TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergeLevel
-        -- TODO: creating the merge should happen in MR.new
-        mergeMaybe <- allocateMaybeTemp reg
-          (Merge.new hfs hbio caching alloc mergeLevel resolve runPaths rs)
-          Merge.abort
-        mr <- case mergeMaybe of
-          Nothing -> error "newMerge: merges can not be empty"
-          Just m  -> allocateTemp reg (MR.new mergePolicy rs m) releaseRef
+        -- TODO: There currently is a resource management bug that happens if an
+        -- exception occurs after calling MR.new. In this case, all changes roll
+        -- back, so some of the runs in rs will live in the Levels structure at
+        -- their original places again. However, we passed their references to
+        -- the MergingRun, which gets aborted, releasing the run references.
+        -- Instead of passing the original references into newMerge, we have to
+        -- duplicate the ones that previously existed in the level and then
+        -- freeTemp the original ones. This way, on the happy path the result is
+        -- the same, but if an exception occurs, the original references do not
+        -- get released.
```

Comment:
I agree, we should do exactly what you say here. And that's consistent with what Joris says above: don't adopt references, create new ones and the caller should remove the old ones. I think generally we should try to follow the pattern that the component that allocated the ref then owns it and so also releases it, and we don't transfer ownership.

Reply:
Done.